site/public/js/chatroom.js

```diff
@@ -0,0 +1,176 @@
+/* global csrfToken userId */
+function fetchMessages(chatroomId, my_id) {
+    $.ajax({
+        url: buildCourseUrl(['chat', chatroomId, 'messages']),
+        type: 'GET',
+        dataType: 'json',
+        success: function(responseData) {
+            if (responseData.status === 'success' && Array.isArray(responseData.data)) {
+                responseData.data.forEach(msg => {
+                    let display_name = msg.display_name;
+                    if (msg.user_id === my_id) {
+                        display_name = "me";
+                    }
+                    appendMessage(display_name, msg.role, msg.timestamp, msg.content);
+                });
+                const messages_area = document.querySelector(".messages-area");
+                messages_area.scrollTop = messages_area.scrollHeight;
+            }
+        },
+        error: function() {
+            window.alert('Something went wrong with fetching messages');
+        },
+    });
+}
+
+function sendMessage(chatroomId, userId, displayName, role, content) {
+    $.ajax({
+        url: buildCourseUrl(['chat', chatroomId, 'send']),
+        type: 'POST',
+        data: {
+            'csrf_token': csrfToken,
+            'user_id': userId,
+            'display_name': displayName,
+            'role': role,
+            'content': content
+        },
+        error: function() {
+            window.alert('Something went wrong with storing message');
+        },
+    })
+    window.socketClient.send({'type': "chat_message", 'content': content, 'user_id': userId, 'display_name': displayName, 'role': role, 'timestamp': new Date(Date.now()).toLocaleString()})
+    appendMessage("me", role, null, content);
+}
+
+function appendMessage(displayName, role, ts, content) {
+    let timestamp = ts;
+    if (!timestamp) {
+        timestamp = new Date(Date.now()).toLocaleString('en-us',  { year:"numeric", month:"short", day:"numeric", hour:"numeric", minute:"numeric"});
+    }
+    else {
+        timestamp = new Date(ts).toLocaleString('en-us', { year:"numeric", month:"short", day:"numeric", hour:"numeric", minute:"numeric"});
+    }
+
+    let display_name = displayName;
+    if (role && role !== 'student' && display_name !== 'me') {
+        display_name = `${displayName}[instructor]`;
+    }
+
+    const messages_area = document.querySelector('.messages-area');
+    const message = document.createElement('div');
+    message.classList.add('message-container');
+    if (role === "instructor") {
+        message.classList.add('admin-message')
+    }
+    message.innerHTML = `
+        <div class="message-header">
+            <span class="sender-name">${display_name}</span>
+            <span class="timestamp">${timestamp}</span>
+        </div>
+        <div class="message-content">
+            ${content}
+        </div>
+    `;
```

Comment:
This is dangerous to do because you are using .innerHTML. This could be subject to an XSS attack. Any variable that contains data coming from the user should be sanitized. Easiest way to do this is to create any needed elements and then insert user data with .innerText or .text() if using jQuery.